app/extlinux_control.py

```diff
@@ -542,18 +606,86 @@ def _is_switching_boot(self) -> bool:
 
         # evidence 3: slot in use
         # the slot_in_use file should have the same slot as current slot
-        _is_slot_in_use = self.load_slot_in_use_file() == Nvbootctrl.get_current_slot()
+        _is_slot_in_use = (
+            self.load_slot_in_use_file() == self._boot_control.get_current_slot()
+        )
 
         logger.debug(
             f"[checking result] nvboot: {_nvboot_res}, ota_status: {_ota_status_res}, slot_in_use: {_is_slot_in_use}"
         )
         return _nvboot_res and _ota_status_res and _is_slot_in_use
 
+    def _populate_extlinux_cfg_to_separate_bootdev(self):
+        if not self._boot_control.is_external_rootfs_enabled():
+            return
+
+        src: Path = self._mount_point / Path(cfg.EXTLINUX_FILE).relative_to("/")
+        target_dir: Path = (
+            self._standby_boot_mount_point
+            / Path(cfg.EXTLINUX_FILE).relative_to("/").parent
+        )
+
+        # ensure target dir's present
+        target_dir.mkdir(parents=True, exist_ok=True)
+
+        if src.is_file():
+            shutil.copy(str(src), target_dir)
+        else:
+            raise OtaErrorUnrecoverable(
+                "extlinux.cfg on boot partition and/or standby slot not found"
+            )
+
+    def _populate_boot_folder_to_separate_bootdev(self):
+        if not self._boot_control.is_external_rootfs_enabled():
+            return
+
+        # copy the /boot folder from standby slot to boot dev unconditionally
+        tmp_dir = self._standby_boot_mount_point / ".tmp_boot"
+        dst_dir = self._standby_boot_mount_point / "boot"
+        dst_dir.mkdir(parents=True, exist_ok=True)  # ensure dst
+        # prepare temporary files
+        if tmp_dir.is_dir():  # cleanup previous temp
+            shutil.rmtree(str(tmp_dir), ignore_errors=True)
+
+        shutil.copytree(
+            str(self._mount_point / "boot"), tmp_dir, symlinks=True, dirs_exist_ok=True
+        )
+
+        src_files = tmp_dir.glob("**/*")
```

Comment:
@Bodong-Yang 
I think coping to tmp_dir is not required since this function is handling multiple files, so the operation is not atomic.
And I prefer using `cp -d -r -p` to copy tree. Using cp, new code is not required and we don't need to debug, unit test (in my opinion)

Reply:
The problem is we have to delete the old `/boot` before we can `cp` the new one to the boot partition.
It seems there is no way to do the `delete and copy` operation atomically. So I choose to `first copy, and then rename` method. `Path.rename` and `Path.replace` is atomic(when src and dst are under the same partition) so we can ensure that unexpected reboot will not result in missing files.